torch/nn/quantized/dynamic/modules/linear.py

```diff
@@ -0,0 +1,72 @@
+from __future__ import absolute_import, division, print_function, unicode_literals
+import torch
+from ....modules.linear import Linear as NNLinear
+import torch.nn.quantized as nnq
+
+class Linear(nnq.Linear):
+    r"""
+    A dynamic quantized linear module with quantized tensor as inputs
+    and outputs.
+    We adopt the same interface as `torch.nn.Linear`, please see
+    https://pytorch.org/docs/stable/nn.html#torch.nn.Linear for documentation.
+
+    Similar to `torch.nn.Linear`, attributes will be randomly initialized at
+        module creation time and will be overwritten later
+
+    Attributes:
+        weight: the non-learnable quantized weights of the
+                module which are of shape :math:`(\text{out\_features}, \text{in\_features})`.
+        bias:   the non-learnable bias of the module of shape :math:`(\text{out\_features})`.
+                If :attr:`bias` is ``True``, the values are initialized to zero.
+        scale: `scale` parameter of weight Quantized Tensor, type: double
+        zero_point: `zero_point` parameter for weight Quantized Tensor, type: long
+
+    Examples::
+
+        >>> m = nn.quantized.dynamic.Linear(20, 30)
+        >>> input = torch.randn(128, 20)
+        >>> output = m(input)
+        >>> print(output.size())
+        torch.Size([128, 30])
+    """
+    def __init__(self, in_features, out_features, bias=True):
+        super(Linear, self).__init__(in_features, out_features, bias)
+        if bias:
+            del self.bias
+            bias = torch.Tensor(out_features)
+            self.register_buffer('bias', bias)
+        else:
+            self.register_buffer('bias', None)
+        del self.scale
+        del self.zero_point
+
+    def forward(self, x):
+        # Note that we can handle self.bias == None case.
+        Y = torch.ops.quantized.fbgemm_linear_dynamic(
+            x, self._packed_weight,
+            self.bias)
+        return Y.to(x.dtype)
+
+    # TODO: support initializing from quantization parameters when Quantizer is
+    # exposed in python
+    @classmethod
+    def from_float(cls, mod):
+        r"""Create a dynamic quantized module from a float module or qparams_dict
+
+            Args: `mod` a float module, either produced by torch.quantization utilities
+            or directly from user
+        """
+        assert type(mod) == NNLinear, 'nnq.Linear.from_float only works for nn.Linear'
+        # assert type(mod) == cls.__FLOAT_MODULE, ' nnq.' + cls.__name__ + '.from_float only works for ' + \
+        #                     cls.__FLOAT_MODULE.__name__
+        assert hasattr(mod, 'qconfig'), 'Input float module must have qconfig defined'
+        assert hasattr(mod, 'observer'), 'Input float module must have observer attached'
+        weight_observer = mod.qconfig.weight()
```

Comment:
given that dynamic quantization is so simple to use, I'd allow also to not have qconfig specified in which case you use the default one. Also, maybe also allow to override qconfig (or rather weight_observer) as an input argument?

Reply:
cc @jerryzh168 . After https://github.com/pytorch/pytorch/pull/23561, for dynamic quantization, we don't need to insert the observer.

Reply:
what I meant is to have the signature `def from_float(cls, mod, weight_observer=default_weight_observer)`. That way I don't even need to do prepare for the simple case - I can just pass in `nnd.Linear(mod)` like in https://github.com/pytorch/pytorch/blob/master/torch/jit/quantized.py#L13

Reply:
This is doable. However, it caused the circular import issues in Python. Basically, when we use `default_weight_observer` in `torch/nn/quantized/dynamic/modules/linear.py`, we need to import `torch.quantization` in `torch.quantized.dynamic` namespace. In the meanwhile, we also need to import `torch.quantized.dynamic` in `torch.quantization` as we need to convert the modules. The error message is like the following:
```
stderr:
Traceback (most recent call last):
  File "<string>", line 38, in <module>
  File "<string>", line 36, in __run
  File "/usr/local/fbcode/platform007/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/local/fbcode/platform007/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/caffe2/test/quantized#binary,link-tree/__fb_test_main__.py", line 845, in <module>
    sys.exit(main(sys.argv))
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/caffe2/test/quantized#binary,link-tree/__fb_test_main__.py", line 841, in main
    return MainProgram(sys.argv).run()
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/caffe2/test/quantized#binary,link-tree/__fb_test_main__.py", line 695, in run
    test_suite = self.load_tests()
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/caffe2/test/quantized#binary,link-tree/__fb_test_main__.py", line 678, in load_tests
    suite = loader.load_all()
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/caffe2/test/quantized#binary,link-tree/__fb_test_main__.py", line 467, in load_all
    __import__(module_name, level=0)
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/caffe2/test/quantized#binary,link-tree/test_nn_quantized.py", line 8, in <module>
    import torch.nn.quantized.dynamic as nnqd
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/caffe2/test/quantized#binary,link-tree/torch/nn/quantized/dynamic/__init__.py", line 4, in <module>
    from .modules import *  # noqa: F401
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/caffe2/test/quantized#binary,link-tree/torch/nn/quantized/dynamic/modules/__init__.py", line 3, in <module>
    from .linear import Linear
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/caffe2/test/quantized#binary,link-tree/torch/nn/quantized/dynamic/modules/linear.py", line 5, in <module>
    from torch.quantization.observer import default_weight_observer
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/caffe2/test/quantized#binary,link-tree/torch/quantization/__init__.py", line 2, in <module>
    from .quantize import *  # noqa: F401
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/caffe2/test/quantized#binary,link-tree/torch/quantization/quantize.py", line 4, in <module>
    import torch.nn.quantized.dynamic as nnqd
AttributeError: module 'torch.nn.quantized' has no attribute 'dynamic'
```
Working on fixing this. Let me know if you have any insights.

Reply:
Currently the observer class and qconfig class are in “torch.quantization” namespace. If we need the observer and qconfig in “torch.nn.quantize.dynamic”, we need to import “torch.quantization” (Check one example: https://github.com/pytorch/pytorch/pull/23128#discussion_r309528932). In the meanwhile, in the “torch.quantization” namespace, we need to import “torch.nn.quantize.dynamic”, because we need to convert the modules and specify the “torch.nn.quantize.dynamic.Linear” in the mapping. 

The possible solution is to move the “QConfig” and “observer” out of torch.quantization namespace.

Ping @dzhulgakov , @raghuramank100 , @jerryzh168 for suggestions.

Reply:
I think the standard recommendation is to move observer and qconfig into a separate submodule (hidden, just import everything in __init__ as well), then torch.nn.quantize.dynamic can import that submodule instead of the main one

Reply:
I would like to resolve in another PR, as I need to add the `default_weight_observer` for both `torch.quantized.dynamic.Linear` and `torch.quantized.Linear`. In addition, I need to move the observer and qconfig into a separate submodule.